server/src/main/java/org/elasticsearch/index/mapper/TsdbIdFieldMapper.java

```diff
@@ -0,0 +1,188 @@
+ * stored, but we need to keep it so that its FieldType can be used to generate
+ * queries.
+ */
+public class TsdbIdFieldMapper extends IdFieldMapper {
+    public static class Defaults {
+
+        public static final FieldType FIELD_TYPE = new FieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.setStored(true);  // TODO reconstruct the id on fetch from tsid and timestamp
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.freeze();
+        }
+    }
+
+    public static final TsdbIdFieldMapper INSTANCE = new TsdbIdFieldMapper();
+
+    public static final TypeParser PARSER = new FixedTypeParser(MappingParserContext::idFieldMapper);
+
+    static final class IdFieldType extends TermBasedFieldType {
+        IdFieldType() {
+            super(NAME, true, true, false, TextSearchInfo.SIMPLE_MATCH_ONLY, Collections.emptyMap());
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public boolean isSearchable() {
+            // The _id field is always searchable.
+            return true;
+        }
+
+        @Override
+        public ValueFetcher valueFetcher(SearchExecutionContext context, String format) {
+            return new StoredValueFetcher(context.lookup(), NAME);
+        }
+
+        @Override
+        public Query termQuery(Object value, SearchExecutionContext context) {
+            return termsQuery(Arrays.asList(value), context);
+        }
+
+        @Override
+        public Query existsQuery(SearchExecutionContext context) {
+            return new MatchAllDocsQuery();
+        }
+
+        @Override
+        public Query termsQuery(Collection<?> values, SearchExecutionContext context) {
+            failIfNotIndexed();
+            BytesRef[] bytesRefs = values.stream().map(v -> {
+                Object idObject = v;
+                if (idObject instanceof BytesRef) {
+                    idObject = ((BytesRef) idObject).utf8ToString();
+                }
+                return Uid.encodeId(idObject.toString());
+            }).toArray(BytesRef[]::new);
+            return new TermInSetQuery(name(), bytesRefs);
+        }
+    }
+
+    private TsdbIdFieldMapper() {
+        super(new IdFieldType(), Lucene.KEYWORD_ANALYZER);
+    }
+
+    @Override
+    public void postParse(DocumentParserContext context) throws IOException {
+        int routingHash = 0;
+        int nonRoutingHash = 0;
+        for (Map.Entry<BytesRef, DimensionInfo> entry : context.doc().getDimensions().entrySet()) {
+            int thisHash = hash(entry.getKey()) ^ hash(entry.getValue().tsidBytes());
+            if (entry.getValue().isRoutingDimension()) {
+                routingHash = 31 * routingHash + thisHash;
+            } else {
+                nonRoutingHash = 31 * nonRoutingHash + thisHash;
```

Comment:
I talked with @henningandersen about this one and he linked me to https://preshing.com/20110504/hash-collision-probabilities/ . We had some brainstorming. In the worst case the `nonRoutingHash` and the `routingHash` are entirely correlated -imagine we route on `data_center` and use `ip` a the only other dimension. It's not great, but it could happen. In that case the only things saving us from unexpected `_id` collisions is the `nonRoutingHash` and the timer. And some folks are going to have garbage resolution timers.

So in the worst case the odds of an `_id` collision are entirely bases on the odds of collision on `nonRoutingHash`. And, that's a birthday problem. Or, so says the link Henning shared. I buy that. That link has a handy table. For the odds of any two hashes colliding, assuming this hashing algorithm is perfect, depend on the number of unique tsids and the number of bits in the hash. For the 32 bit hash I'm using here it takes 30,000 to get about 1:10 odds of any two colliding. That seems bad.

So I'll have to think more here. A 64 bit hash would be a lot better - 1:10 odds takes 1.97 billion unique tsid. Maybe that'd be enough.

Reply:
Instead of a hash, it'd be nice if we could use a lookup table to sequentially assign the tsids. Lucene already *has* a lookup table for global ordinals, but it's global ordinals, an expensive query time thing. It has isn't really for this. It's in the wrong place and has the wrong constraints. We'd have to build something else. And that'd need replication semantics and all that jazz. So, possible, but I think the path of least resistance is to go with a 64 bit hash and live with the rare chance of collisions for a while. I'd love to replace it with something fancier, but that's a bigger project I think.